File: app/src/androidMain/kotlin/com/thejawnpaul/gptinvestor/core/firebase/UserImpl.kt
package com.thejawnpaul.gptinvestor.core.firebase

import com.google.firebase.auth.FirebaseUser
import kotlinx.coroutines.tasks.await

class UserImpl(private val android: FirebaseUser) : IUser {
    override val uid: String
        get() = android.uid
    override val email: String?
        get() = android.email
    override val displayName: String?
        get() = android.displayName
    override val providerId: String
        get() = android.providerId

    override suspend fun delete() {
        android.delete().await()
    }
}

fun FirebaseUser.toUser(): IUser = UserImpl(this)

File: iosApp/iosApp/Firebase/UserImpl.swift
//
//  UserImpl.swift
//  iosApp
//
//  Created by ABDULKARIM ABDULRAHMAN on 11/09/2025.
//  Copyright © 2025 orgName. All rights reserved.
//
import GPT_Investor
import FirebaseAuth

class UserImpl: IUser {
    var user: FirebaseAuth.User!
    
    init(user: FirebaseAuth.User!) {
        self.user = user
    }
    
    func delete(completionHandler: @escaping ((any Error)?) -> Void) {
        self.delete {_ in
            completionHandler(nil)
        }
    }
    
    var displayName: String? {
        return user.displayName
    }
    
    var email: String? {
        return user.email
    }
    
    var providerId: String {
        return user.providerID
    }

    var uid: String {
        return user.uid
    }
}

extension User {
    func toUser() -> IUser {
        UserImpl(user: self)
    }
}
